front/src/applications/operationalStudies/components/MacroEditor/NGE.tsx

```diff
@@ -22,29 +33,33 @@ const frameSrc = `
 </html>
 `;
 
-const NGE = () => {
+const NGE = ({ dto }: { dto?: NetzgrafikDto }) => {
   const frameRef = useRef<HTMLIFrameElement>(null);
+  const dispatch = useAppDispatch();
+
+  const [ngeRootElement, setNgeRootElement] = useState<NGEElement | null>(null);
+
+  const { getTimetableID } = useOsrdConfSelectors();
+  const timeTableId = useSelector(getTimetableID);
```

Comment:
Up until now, I've designed the `NGE` component to be completely self-contained. That is, it's not tied to anything OSRD-specific. The goal is to have a clear separation of concerns: keep the OSRD-specific logic outside of the component, and have the component only define an API to manage NGE without anything else. That way the low-level NGE details (iframe management, typing etc) are isolated away from the rest of the code.

To retain this design, we'd need to remove this and instead expose a `onOperation` prop that callers can set to a callback taking an `Operation` argument.

Do you think it's better to keep this design, or do you think there are good reason to diverge from it?

Reply:
No you're right to want to separate NGE and OSRD.

Reply:
Here's what I had in mind:

```ts
// types.ts
export type NGEEvent = NGECreateEvent | NGEDeleteEvent | NGEUpdateEvent;

// NGE.tsx
type NGEProps = {
  dto?: NetzgrafikDto;
  onOperation?: (event: NGEEvent);
};
// ...
ngeRoot.addEventListener('operation', async (event: Event) => {
  const customEvent = event as CustomEvent;
  const op = customEvent.detail as NGEEvent;
  if (onOperation) onOperation(op);
});

// ScenarioV2.tsx
const handleNGEOperation = (op: NGEEvent) =>
  ngeOperations(op, dispatch, timeTableId);
// ...
<NGE dto={ngeDto} onOperation={handlNGEOperation} />
```